resim/msg/resim_log_from_ros2_test.cc

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 ReSim, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#include "resim/msg/resim_log_from_ros2.hh"
+
+#include <gtest/gtest.h>
+
+#include <cstring>
+#include <filesystem>
+#include <map>
+#include <mcap/reader.hpp>
+#include <rclcpp/serialization.hpp>
+#include <rclcpp/serialized_message.hpp>
+#include <rosbag2_cpp/writer.hpp>
+#include <std_msgs/msg/byte.hpp>
+#include <string>
+#include <tuple>
+
+#include "resim/assert/assert.hh"
+#include "resim/msg/fuzz_helpers.hh"
+#include "resim/msg/odometry_from_ros2.hh"
+#include "resim/msg/transform_from_ros2.hh"
+#include "resim/testing/fuzz_helpers.hh"
+#include "resim/testing/test_directory.hh"
+#include "resim/time/timestamp.hh"
+#include "resim/utils/proto/dependency_file_descriptor_set.hh"
+#include "resim/utils/tuple_utils.hh"
+#include "rosbag2_storage/storage_options.hpp"
+
+namespace resim::msg {
+
+namespace {
+
+// A struct template storing the type and name of a given channel.
+template <typename T>
+struct Channel {
+  using type = T;
+  const char *name;
+};
+
+// The set of channels we would like to test.
+constexpr auto TEST_CHANNELS = std::make_tuple(
+    Channel<TransformArray>{"/transforms"},
+    Channel<TransformArray>{"/transforms_also"},
+    Channel<Odometry>{"/odom"});
+
+// Nested maps storing Channel -> (Timestamp -> Message)
+using MessageMap = std::map<
+    std::string,
+    std::map<mcap::Timestamp, std::unique_ptr<google::protobuf::Message>>>;
+
+constexpr int MESSAGES_PER_CHANNEL = 50;
+
+// Helper function to make MESSAGES_PER_CHANNEL test messages on each channel.
+MessageMap make_test_messages() {
+  constexpr std::size_t SEED = 913U;
+  std::mt19937 rng{SEED};
+
+  MessageMap result;
+  for_each_in_tuple(
+      [&rng, &result](auto channel) {
+        using MessageType = typename decltype(channel)::type;
+        for (int ii = 0; ii < MESSAGES_PER_CHANNEL; ++ii) {
+          const mcap::Timestamp stamp =
+              resim::random_element<uint32_t>(InOut{rng});
+          REASSERT(result[channel.name]
+                       .emplace(
+                           stamp,
+                           std::make_unique<MessageType>(
+                               random_element<MessageType>(InOut{rng})))
+                       .second);
+        }
+        // Need to return something even though we don't use it.
+        return true;
```

Comment:
Discussed IRL

Reply:
I'm going to leave this as tech debt for now, since it's a bit ugly to work around quickly.